File: implementation/01 Fundamentals/Concurreny Basics/Thread Pool/java/Main.java
import java.util.Scanner;
import java.util.concurrent.TimeUnit;

public class Main {
    public static void main(String[] args) {
        System.out.println("=== Interactive Java Thread Pool Demo ===");
        System.out.println("Initializing pool with Size: 2, Queue Capacity: 3, Rejection Policy: AbortPolicy");

        // Create the thread pool
        FixedThreadPool pool = FixedThreadPool.builder()
                .poolSize(2)
                .queueCapacity(3)
                .rejectionPolicy(new RejectionPolicy.AbortPolicy())
                .build();

        Scanner scanner = new Scanner(System.in);
        System.out.println("\nCommands:");
        System.out.println("  submit <id> <delay_ms> - Submit a task that sleeps for delay_ms");
        System.out.println("  status                - View pool stats (active workers, queue size)");
        System.out.println("  shutdown              - Gracefully shut down and wait");
        System.out.println("  shutdownnow           - Immediately shut down");
        System.out.println("  help                  - Print this menu");
        System.out.println("  exit                  - Exit the program");

        while (true) {
            System.out.print("\n> ");
            if (!scanner.hasNextLine()) {
                break;
            }
            String input = scanner.nextLine().trim();
            if (input.isEmpty()) {
                continue;
            }

            String[] parts = input.split("\\s+");
            String cmd = parts[0].toLowerCase();

            switch (cmd) {
                case "submit":
                    if (parts.length < 3) {
                        System.out.println("Error: submit requires <id> and <delay_ms>");
                        continue;
                    }
                    String id = parts[1];
                    int delayMs;
                    try {
                        delayMs = Integer.parseInt(parts[2]);
                    } catch (NumberFormatException e) {
                        System.out.println("Error: invalid delay_ms");
                        continue;
                    }

                    Runnable task = () -> {
                        System.out.printf("[Task %s] Starting (will run for %dms)\n", id, delayMs);
                        try {
                            Thread.sleep(delayMs);
                        } catch (InterruptedException e) {
                            System.out.printf("[Task %s] Interrupted!\n", id);
                            Thread.currentThread().interrupt();
                            return;
                        }
                        System.out.printf("[Task %s] Completed\n", id);
                    };

                    try {
                        pool.submit(task);
                        System.out.printf("Submit Accepted: Task %s\n", id);
                    } catch (RejectedTaskException e) {
                        System.out.printf("Submit Rejected: %s\n", e.getMessage());
                    } catch (NullPointerException e) {
                        System.out.println("Error: task is null");
                    }
                    break;

                case "status":
                    System.out.println("Pool Status:");
                    System.out.printf("  Active Workers: %d\n", pool.getActiveWorkerCount());
                    System.out.printf("  Queue Size:     %d\n", pool.getQueueSize());
                    System.out.printf("  Is Shutdown:    %b\n", pool.isShutdown());
                    System.out.printf("  Is Terminated:  %b\n", pool.isTerminated());
                    break;

                case "shutdown":
                    System.out.println("Initiating graceful shutdown...");
                    pool.shutdown();
                    System.out.println("Waiting for workers to finish (timeout 30s)...");
                    try {
                        boolean clean = pool.awaitTermination(30, TimeUnit.SECONDS);
                        System.out.printf("Pool terminated cleanly: %b\n", clean);
                    } catch (InterruptedException e) {
                        System.out.println("Interrupted waiting for termination.");
                        Thread.currentThread().interrupt();
                    }
                    return;

                case "shutdownnow":
                    System.out.println("Initiating immediate shutdown...");
                    pool.shutdownNow();
                    System.out.println("Waiting for workers to exit (timeout 5s)...");
                    try {
                        boolean clean = pool.awaitTermination(5, TimeUnit.SECONDS);
                        System.out.printf("Pool terminated cleanly: %b\n", clean);
                    } catch (InterruptedException e) {
                        System.out.println("Interrupted waiting for termination.");
                        Thread.currentThread().interrupt();
                    }
                    return;

                case "help":
                    System.out.println("Commands:");
                    System.out.println("  submit <id> <delay_ms> - Submit a task that sleeps for delay_ms");
                    System.out.println("  status                - View pool stats (active workers, queue size)");
                    System.out.println("  shutdown              - Gracefully shut down and wait");
                    System.out.println("  shutdownnow           - Immediately shut down");
                    System.out.println("  help                  - Print this menu");
                    System.out.println("  exit                  - Exit the program");
                    break;

                case "exit":
                    if (!pool.isShutdown()) {
                        pool.shutdownNow();
                        try {
                            pool.awaitTermination(5, TimeUnit.SECONDS);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    }
                    return;

                default:
                    System.out.printf("Unknown command: %s. Type 'help' for commands.\n", cmd);
                    break;
            }
        }
    }
}

File: implementation/01 Fundamentals/Concurreny Basics/Thread Pool/go/cmd/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"threadpool"
	"time"
)

func main() {
	fmt.Println("=== Interactive Go Thread Pool Demo ===")
	fmt.Println("Initializing pool with Size: 2, Queue Capacity: 3, Rejection Policy: AbortPolicy")
	
	// Create the thread pool
	pool := threadpool.New(threadpool.Config{
		PoolSize:        2,
		QueueCapacity:   3,
		RejectionPolicy: threadpool.AbortPolicy,
	})

	reader := bufio.NewReader(os.Stdin)
	fmt.Println("\nCommands:")
	fmt.Println("  submit <id> <delay_ms> - Submit a task that sleeps for delay_ms")
	fmt.Println("  status                - View pool stats (active workers, queue size)")
	fmt.Println("  shutdown              - Gracefully shut down and wait")
	fmt.Println("  shutdownnow           - Immediately shut down")
	fmt.Println("  help                  - Print this menu")
	fmt.Println("  exit                  - Exit the program")

	for {
		fmt.Print("\n> ")
		input, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		input = strings.TrimSpace(input)
		if input == "" {
			continue
		}

		parts := strings.Split(input, " ")
		cmd := strings.ToLower(parts[0])

		switch cmd {
		case "submit":
			if len(parts) < 3 {
				fmt.Println("Error: submit requires <id> and <delay_ms>")
				continue
			}
			id := parts[1]
			delayMs, err := strconv.Atoi(parts[2])
			if err != nil {
				fmt.Println("Error: invalid delay_ms")
				continue
			}

			task := func() {
				fmt.Printf("[Task %s] Starting (will run for %dms)\n", id, delayMs)
				time.Sleep(time.Duration(delayMs) * time.Millisecond)
				fmt.Printf("[Task %s] Completed\n", id)
			}

			err = pool.Submit(task)
			if err != nil {
				fmt.Printf("Submit Rejected: %v\n", err)
			} else {
				fmt.Printf("Submit Accepted: Task %s\n", id)
			}

		case "status":
			fmt.Printf("Pool Status:\n")
			fmt.Printf("  Active Workers: %d\n", pool.ActiveWorkerCount())
			fmt.Printf("  Queue Size:     %d\n", pool.QueueSize())
			fmt.Printf("  Is Shutdown:    %v\n", pool.IsShutdown())

		case "shutdown":
			fmt.Println("Initiating graceful shutdown...")
			pool.Shutdown()
			fmt.Println("Waiting for workers to finish...")
			pool.Wait()
			fmt.Println("Pool fully terminated.")
			return

		case "shutdownnow":
			fmt.Println("Initiating immediate shutdown...")
			pool.ShutdownNow()
			fmt.Println("Waiting for workers to exit...")
			pool.Wait()
			fmt.Println("Pool terminated.")
			return

		case "help":
			fmt.Println("Commands:")
			fmt.Println("  submit <id> <delay_ms> - Submit a task that sleeps for delay_ms")
			fmt.Println("  status                - View pool stats (active workers, queue size)")
			fmt.Println("  shutdown              - Gracefully shut down and wait")
			fmt.Println("  shutdownnow           - Immediately shut down")
			fmt.Println("  help                  - Print this menu")
			fmt.Println("  exit                  - Exit the program")

		case "exit":
			if !pool.IsShutdown() {
				pool.ShutdownNow()
				pool.Wait()
			}
			return

		default:
			fmt.Printf("Unknown command: %s. Type 'help' for commands.\n", cmd)
		}
	}
}
